ssh/keygen.go

```diff
@@ -0,0 +1,119 @@
+package ssh
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os/exec"
+	"path"
+	"strconv"
+
+	"github.com/spf13/pflag"
+)
+
+// OptionalValue is an extension of pflag.Value that remembers if it has been
+// set.
+type OptionalValue interface {
+	pflag.Value
+	Specified() bool
+}
+
+// KeyBitsValue is an OptionalValue allowing specification of the -b argument
+// to ssh-keygen.
+type KeyBitsValue struct {
+	specified bool
+	keyBits   uint64
+}
+
+func (kbv *KeyBitsValue) String() string {
+	return strconv.FormatUint(kbv.keyBits, 10)
+}
+
+func (kbv *KeyBitsValue) Set(s string) error {
+	v, err := strconv.ParseUint(s, 0, 64)
+	if err != nil {
+		return err
+	}
+
+	kbv.keyBits = v
+	kbv.specified = true
+
+	return nil
+}
+
+func (kbv *KeyBitsValue) Type() string {
+	return "uint64"
+}
+
+func (kbv *KeyBitsValue) Specified() bool {
+	return kbv.specified
+}
+
+// KeyTypeValue is an OptionalValue allowing specification of the -t argument
+// to ssh-keygen.
+type KeyTypeValue struct {
+	specified bool
+	keyType   string
+}
+
+func (ktv *KeyTypeValue) String() string {
+	return ktv.keyType
+}
+
+func (ktv *KeyTypeValue) Set(s string) error {
+	if len(s) > 0 {
+		ktv.keyType = s
+		ktv.specified = true
+	}
+	return nil
+}
+
+func (ktv *KeyTypeValue) Type() string {
+	return "string"
+}
+
+func (ktv *KeyTypeValue) Specified() bool {
+	return ktv.specified
+}
+
+// KeyGen generates a new keypair with ssh-keygen, optionally overriding the
+// default type and size. Each generated keypair is written to a new unique
+// subdirectory of tmpfsPath, which should point to a tmpfs mount as the
+// private key is not encrypted.
+func KeyGen(keyBits, keyType OptionalValue, tmpfsPath string) (privateKeyPath string, privateKey, publicKey []byte, err error) {
+	tempDir, err := ioutil.TempDir(tmpfsPath, "keygen")
+	if err != nil {
+		return "", nil, nil, err
+	}
+
+	privateKeyPath = path.Join(tempDir, "identity")
+	args := []string{"-q", "-N", "", "-f", privateKeyPath}
+	if keyBits.Specified() {
+		args = append(args, "-b", keyBits.String())
+	}
+	if keyType.Specified() {
+		args = append(args, "-t", keyType.String())
+	}
+
+	cmd := exec.Command("ssh-keygen", args...)
+	if err := cmd.Run(); err != nil {
+		return "", nil, nil, err
+	}
+
+	privateKey, err = ioutil.ReadFile(privateKeyPath)
+	if err != nil {
+		return "", nil, nil, err
+	}
+
+	publicKey, err = ioutil.ReadFile(fmt.Sprintf("%s.pub", privateKeyPath))
+	if err != nil {
+		return "", nil, nil, err
+	}
+
+	return privateKeyPath, privateKey, publicKey, nil
+}
+
+// ExtractPublicKey extracts and returns the public key from the specified
+// private key.
+func ExtractPublicKey(privateKeyPath string) (publicKey []byte, err error) {
+	return exec.Command("ssh-keygen", "-y", "-f", privateKeyPath).Output()
```

Comment:
Oh just seen bug comment earlier on. Yes, would be better to generate the key in a more secure way. Good thinking.